rust/lance-encoding/src/encodings/physical/basic.rs

```diff
@@ -167,51 +169,68 @@ struct BasicPageDecoder {
     mode: DataNullStatus,
 }
 
-impl PhysicalPageDecoder for BasicPageDecoder {
-    fn update_capacity(
+impl PrimitivePageDecoder for BasicPageDecoder {
+    fn decode_into(
         &self,
         rows_to_skip: u32,
         num_rows: u32,
-        buffers: &mut [(u64, bool)],
         all_null: &mut bool,
-    ) {
-        // No need to look at the validity decoder to know the dest buffer size since it is boolean
-        buffers[0].0 = arrow_buffer::bit_util::ceil(num_rows as usize, 8) as u64;
-        // The validity buffer is only required if we have some nulls
-        buffers[0].1 = matches!(self.mode, DataNullStatus::Some(_));
-        if let Some(values) = self.mode.values_decoder() {
-            values.update_capacity(rows_to_skip, num_rows, &mut buffers[1..], all_null);
-        } else {
-            *all_null = true;
-        }
-    }
+    ) -> Result<Vec<BytesMut>> {
+        if let Some(_values) = self.mode.values_decoder() {
+            match &self.mode {
```

Comment:
I think you should match on mode first.  Then you can just unwrap `self.mode.values_decoder()` when you need it.  The end result should be more readable.